src/state_machine.rs

```diff
@@ -6,35 +6,46 @@
 // KIND, either express or implied. Please review the Licences for the specific language governing
 // permissions and limitations relating to use of the SAFE Network Software.
 
-use crate::action::Action;
-use crate::id::{FullId, PublicId};
+use crate::{
+    action::Action,
+    chain::GenesisPfxInfo,
+    id::{FullId, PublicId},
+    outbox::EventBox,
+    routing_table::Prefix,
+    states::common::Base,
+    states::{Bootstrapping, Client, Node, ProvingNode, RelocatingNode},
+    timer::Timer,
+    types::RoutingActionSender,
+    xor_name::XorName,
+    BootstrapConfig, {CrustEvent, CrustEventSender, Service, MIN_SECTION_SIZE},
+};
 #[cfg(feature = "mock")]
-use crate::mock_crust;
-use crate::outbox::EventBox;
-#[cfg(feature = "mock")]
-use crate::routing_table::Authority;
-use crate::routing_table::Prefix;
-use crate::states::common::Base;
-#[cfg(feature = "mock")]
-use crate::states::common::Bootstrapped;
-use crate::states::{Bootstrapping, Client, JoiningNode, Node};
-use crate::timer::Timer;
-use crate::types::RoutingActionSender;
-use crate::xor_name::XorName;
-use crate::BootstrapConfig;
-#[cfg(feature = "mock")]
-use crate::Chain;
-use crate::{CrustEvent, CrustEventSender, Service, MIN_SECTION_SIZE};
+use crate::{mock_crust, routing_table::Authority, states::common::Bootstrapped, Chain};
 use log::LogLevel;
 use maidsafe_utilities::event_sender::MaidSafeEventCategory;
 #[cfg(feature = "mock")]
-use std::collections::BTreeMap;
-use std::collections::BTreeSet;
-use std::fmt::{self, Debug, Display, Formatter};
-use std::mem;
-#[cfg(feature = "mock")]
-use std::net::IpAddr;
-use std::sync::mpsc::{self, Receiver, RecvError, Sender, TryRecvError};
+use std::{collections::BTreeMap, net::IpAddr};
+use std::{
+    collections::BTreeSet,
+    fmt::{self, Debug, Display, Formatter},
+    mem,
+    sync::mpsc::{self, Receiver, RecvError, Sender, TryRecvError},
+};
+
+// Execute $expr on the current variant of $self. Execute $term_expr if the current variant is
+// `Terminated`.
+macro_rules! state_dispatch {
+    ($self:expr, $state:pat => $expr:expr, Terminated => $term_expr:expr) => {
+        match $self {
+            State::Bootstrapping($state) => $expr,
+            State::Client($state) => $expr,
+            State::RelocatingNode($state) => $expr,
+            State::ProvingNode($state) => $expr,
+            State::Node($state) => $expr,
+            State::Terminated => $term_expr,
+        }
+    };
+}
```

Comment:
I don't think this would work. The states have each their own type and the current state is known only at runtime, but the closure type needs to be known at compile time.

Reply:
Ah, you're right, of course! So `F` would have to be `Fn(Box<dyn Base>) -> T` or something, but that probably causes other problems…